fri/src/verifier.rs

```diff
@@ -39,48 +49,173 @@ pub struct FriVerifier<
+        }
+
+        self.channel.states.begin_query_phase();
+    }
+
+    pub fn verify_first_layer(&mut self) {
+        let first_fri_step = self.params.fri_step_list[0];
+        let first_layer_queries =
+            second_layer_queries_to_first_layer_queries(&self.query_indices, first_fri_step);
+        let first_layer_result = (self.first_layer_callback)(&first_layer_queries);
+
+        assert_eq!(
+            first_layer_result.len(),
+            first_layer_queries.len(),
+            "Returned number of queries does not match the number sent"
+        );
+        let first_layer_coset_size = 1 << first_fri_step;
+        for i in (0..first_layer_queries.len()).step_by(first_layer_coset_size) {
+            let result = apply_fri_layers(
+                &first_layer_result[i..i + first_layer_coset_size],
+                self.first_eval_point,
+                &self.params,
+                0,
+                first_layer_queries[i] as usize,
+            );
+            self.query_results.push(result);
+        }
+    }
+
+    pub fn verify_inner_layers(&mut self) {
+        let first_fri_step = self.params.fri_step_list[0];
+        let mut basis_index = 0;
+
+        for i in 0..self.n_layers - 1 {
+            let cur_fri_step = self.params.fri_step_list[i + 1];
+            basis_index += self.params.fri_step_list[i];
+
+            let (layer_data_queries, layer_integrity_queries) =
+                next_layer_data_and_integrity_queries(&self.params, &self.query_indices, i + 1);
+
+            let mut to_verify = if let Some(test_data) = self.to_verify_test.as_ref() {
+                test_data[i].clone()
+            } else {
+                self.table_verifiers[i]
+                    .query(
+                        &mut self.channel,
+                        &layer_data_queries,
+                        &layer_integrity_queries,
+                    )
+                    .unwrap()
+            };
+
+            for j in 0..self.query_results.len() {
+                let query_index = self.query_indices[j] >> (basis_index - first_fri_step);
+                let query_loc = get_table_prover_row_col(query_index, cur_fri_step);
+                to_verify.insert(query_loc, self.query_results[j]);
+            }
+
+            let eval_point = self.eval_points[i];
+            for j in 0..self.query_results.len() {
+                let coset_size = 1 << cur_fri_step;
+                let mut coset_elements: Vec<F> = Vec::with_capacity(coset_size);
+                let coset_start = get_table_prover_row(
+                    self.query_indices[j] >> (basis_index - first_fri_step),
+                    cur_fri_step,
+                );
+
+                for k in 0..coset_size {
+                    coset_elements.push(*to_verify.get(&RowCol::new(coset_start, k)).unwrap());
+                }
+
+                self.query_results[j] = apply_fri_layers(
```

Comment:
Does this not refold everything for every level? i.e. n^2 behavior in the number of levels (n)? n is small (log(N)) but still. i.e. in the first call you fold 1 level, in the second you fold 2 levels, ...